lib/Mojo/IOLoop/Stream/HTTPClient.pm

```diff
@@ -0,0 +1,202 @@
+}
+
+sub _finish {
+  my ($self, $close) = @_;
+
+  # Remove request timeout and finish transaction
+  $self->reactor->remove($self->{req_timeout}) if $self->{req_timeout};
+  return ++$self->{closing} && $self->close unless my $tx = delete $self->{tx};
+
+  # Premature connection close
+  my $res = $tx->res->finish;
+  if ($close && !$res->code && !$res->error) {
+    $res->error({message => 'Premature connection close'});
+  }
+
+  # Upgrade connection to WebSocket
+  if (my $ws = $self->_upgrade($tx)) {
+    $self->emit(upgrade => $ws);
+    return $ws->client_read($ws->handshake->res->content->leftovers);
+  }
+
+  ++$self->{closing} && $self->close_gracefully
+    if $tx->error || !$tx->keep_alive;
+  $res->error({message => $res->message, code => $res->code}) if $res->is_error;
+  $tx->closed;
+}
+
+sub _read_content {
+  my ($self, $chunk) = @_;
+
+  # Corrupted connection
+  return $self->close unless my $tx = $self->{tx};
+
+  warn term_escape "-- Client <<< Server (@{[_url($tx)]})\n$chunk\n" if DEBUG;
+  $tx->client_read($chunk);
+  $self->_finish if $tx->is_finished;
+}
+
+sub _upgrade {
+  my ($self, $tx) = @_;
+  my $code = $tx->res->code // 0;
+  return undef unless $tx->req->is_handshake && $code == 101;
+  my $ws = Mojo::Transaction::WebSocket->new(handshake => $tx, masked => 1);
+  return challenge($ws) ? $ws->established(1) : undef;
+}
+
+sub _url { shift->req->url->to_abs }
+
+sub _write_content {
+  my $self = shift;
+
+  # Protect from resume event recursion
+  return if !(my $tx = $self->{tx}) || $self->{cont_writing};
+  local $self->{cont_writing} = 1;
+  my $chunk = $tx->client_write;
+  warn term_escape "-- Client >>> Server (@{[_url($tx)]})\n$chunk\n" if DEBUG;
+  return unless length $chunk;
+  $self->write($chunk => sub { $_[0]->_write_content });
+}
+
+1;
+
+=encoding utf8
+
+=head1 NAME
+
+Mojo::IOLoop::Stream::HTTPClient - Non-blocking I/O HTTP client
+
+=head1 SYNOPSIS
+
+  use Mojo::IOLoop::Client;
+  use Mojo::IOLoop::Stream::HTTPClient;
+  use Mojo::Transaction::HTTP;
+  
+  # Create transaction
+  my $tx = Mojo::Transaction::HTTP->new;
+  $tx->req->method('GET')->url->parse('http://mojolicious.org');
+  $tx->on(finish => sub {
+      my $tx = shift;
+      say $tx->res->code;
+    });
+  
+  # Create socket connection
+  my $client = Mojo::IOLoop::Client->new;
+  $client->on(
+    connect => sub {
+      my $stream = Mojo::IOLoop::Stream::HTTPClient->new(pop);
+      $stream->start;
+      $stream->process($tx);
+    }
+  );
+  
+  $client->connect(address => 'mojolicious.org', port => 80);
+  $client->reactor->start;
+
+=head1 DESCRIPTION
+
+HTTP client connection based on L<Mojo::IOLoop::Stream>.
+
+=head1 EVENTS
+
+L<Mojo::IOLoop::Stream::HTTPClient> inherits all events from
+L<Mojo::IOLoop::Stream::Client> and can emit the following new ones.
+
+=head2 upgrade
+
+  $stream->on(upgrade => sub {
+    my ($stream, $ws) = @_;
+    ...
+  });
+
+Emitted when connection should be upgraded to WebSocket.
```

Comment:
Maybe `Emitted when the connection should be upgraded to the WebSocket protocol.`?